packages/components/src/number-control/utils.ts

```diff
@@ -0,0 +1,103 @@
+/**
+ * External dependencies
+ */
+import type { KeyboardEvent, MouseEvent } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import { add, subtract, roundClamp } from '../utils/math';
+
+export const computeStep = ( {
+	shiftStep,
+	baseStep,
+	enableShift = false,
+}: {
+	shiftStep: number;
+	baseStep: number;
+	enableShift?: boolean;
+} ) => ( enableShift ? shiftStep * baseStep : baseStep );
+
+type CommonConstrainValueArgs = {
+	value: number;
+	isStepAny: boolean;
+	min: number;
+	max: number;
+	baseStep: number;
+};
+
+/**
+ * Applies clamping based on min/max and step value.
+ * When step is "any" clamp the value, otherwise round and clamp it.
+ *
+ * @param  args
+ * @param  args.value        The value to constrain
+ * @param  args.isStepAny    True is `step==="any"`
+ * @param  args.min          The minimum value that the result can assume
+ * @param  args.max          The maximum value that the result can assume
+ * @param  args.baseStep     The default, base step value
+ * @param  args.stepOverride Override `baseStep` if necessary
+ */
+export const constrainValue = ( {
```

Comment:
I don't know about moving these helpers to their own function. You said elsewhere that it's so that they are pure and testable, but:

1. **On purity**

   Sure, but they're still tightly coupled.

   For example it's odd that the caller would pass `baseStep` and `stepOverride` instead of just `step`. It's also odd that the caller would pass `isStepAny` instead of passing `step` and `constrainValue` figuring that out itself. 

   The functions have the _appearance_ of being decoupled but in reality they're still highly coupled to the component. Why not embrace that and keep them in the component?

2. **On testability**

   Sure, but this PR doesn't include tests 😀

I think either rethink where the "API boundary" of these functions is (e.g. have just `step` and not `baseStep`, `stepOverride`, etc.) and add tests or just go back to having them be inline in the component.

Reply:
I sympathize with wanting to facilitate testing by extracting these but feel that the added parameters hurts their readability/expressiveness. Might it be okay to save for a follow-up?

And yet to noodle on the idea, I wonder if it'd be nicer having `spinValue` and `constrainValue` come back in a return from a higher order function so they could keep their current simpler parameters. Something like:
```javascript
const { constrainValue, spinValue } = getValueHelpers( {
	baseStep,
	isShiftStepEnabled,
	isStepAny,
	min,
	max,
} );
```

Reply:
Thank you for the feedback. I think at this point the easiest thing to do is to revert to having those functions as part of the main file (as it currently is on `trunk`). We can refactor/optimize at a later point, with lower priority

Reply:
Done in https://github.com/WordPress/gutenberg/pull/45982/commits/bc58827055d15e7be1f730ee4ac266794c964ab2